### PR TITLE
Anti Spam

### PR DESCRIPTION
**Pull request by mrkirby153**
Implement anti-spam similar to how rowboat used to do anti-spam.

There are 4 buckets currently that can be configured for anti-spam. Each bucket takes a `count` (The amount of occurrences) and a `period` (The time window over which these occurrences must occur). Once a bucket is "full," a violation is thrown.

### The Buckets
* `max_duplicates` - Tracks identical messages from users
* `max_mentions` - Tracks mentions (channels, role, user) from users
* `max_newlines` - Tracks newlines in messages
* `max_messages` - Tracks individual messages from users

### Actions upon violation
Once a violation occurs, the following punishments are supported:
* `kick` - Kick the user
* `warn` - Warn the user
* `tempban` - Temporarily ban the user (`PUNISHMENT_DURATION` sets the duration in seconds)
* `ban` - Ban the user

In addition, by setting `CLEAN` to true, the offending messages will be deleted (**Note:** Depending on which bucket raised the violation, this may not get all the messages and some manual clean up may be required).

### Exemptions
`EXEMPT_ROLES` contains a list of role ids that are exempt from the spam filter. `EXEMPT_ULSERS` contains a list of user ids that are exempt from the spam filter.

### Additional Notes
* Anti-Spam tracks time using the message's timestamp, not the system time. This attempts to reduce false triggers during an API spike.
* Anti-Spam is disabled by default.

### Things still missing
* Configuration command. Currently, there's no way in the bot to configure anti-spam. I couldn't figure out how to do that

### CODE REVIEW

File: GearBot/Cogs/AntiSpam.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+        super(AntiSpam, self).__init__(bot)
+
+    def get_bucket(self, guild_id, rule_name):
+        cfg = Configuration.get_var(guild_id, "ANTI_SPAM")
+        bucket_cfg = cfg.get(rule_name.upper())
+        if bucket_cfg is None:
+            return None
+        c = bucket_cfg.get("COUNT")
+        p = bucket_cfg.get("PERIOD")
+        return SpamBucket(self.bot.redis_pool, "{}:{}:{}".format(guild_id, rule_name, "{}"), c, p)
```
Comment (Tarek1337) [inline GearBot/Cogs/AntiSpam.py]: Not a big priority, but maybe consider using f-strings here for better readability?

File: GearBot/Cogs/AntiSpam.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+        await check_bucket("max_messages", Translator.translate('spam_max_messages', ctx), 1)
+        await check_bucket("max_newlines", Translator.translate('spam_max_newlines', ctx), len(re.split("\\r\\n|\\r|\\n", ctx.content)))
+        await check_bucket("max_mentions", Translator.translate('spam_max_mentions', ctx), len(re.findall("<@[!&]?\\d+>", ctx.content)))
+        await self.check_duplicates(ctx)
+
+    async def check_duplicates(self, ctx: Message):
+        rule = Configuration.get_var(ctx.guild.id, "ANTI_SPAM")["MAX_DUPLICATES"]
+        spam_bucket = SpamBucket(self.bot.redis_pool,
+                                 "spam:duplicates:{}:{}:{}".format(ctx.guild.id, ctx.author.id, "{}"), rule["COUNT"],
+                                 rule["PERIOD"])
```
Comment (Tarek1337) [inline GearBot/Cogs/AntiSpam.py]: Same here, f-string maybe?

File: GearBot/Cogs/AntiSpam.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+                        # Ensure we only delete 100 at a time. Probably not necessary but you never know with people
+                        for group in divide_msgs(msgs):
+                            await guild_chan.delete_messages(group)
+
+    @staticmethod
+    def is_exempt(guild_id, member: Member):
+        config = Configuration.get_var(guild_id, "ANTI_SPAM")
+        for role in member.roles:
+            if str(role.id) in config["EXEMPT_ROLES"]:
```
Comment (Tarek1337) [inline GearBot/Cogs/AntiSpam.py]: why are you converting to a string?

File: GearBot/Util/Configuration.py [hunk 4]
```diff
@@ -138,14 +138,40 @@ def v9(config):
...
+        },
+        "MAX_MESSAGES": {
+            "COUNT": 7,
+            "PERIOD": 10
+        },
+        "MAX_NEWLINES": {
+            "COUNT": 20,
+            "PERIOD": 60
+        },
+        "PUNISHMENT": "none",
```
Comment (Tarek1337) [inline GearBot/Util/Configuration.py]: Wouldn't ``null`` as default value be more appropriate? So Python can return None and not a string just saying ``none``.

File: config/template.json [hunk 5]
```diff
@@ -30,5 +30,29 @@
...
+    },
+    "MAX_MESSAGES": {
+      "COUNT": 7,
+      "PERIOD": 10
+    },
+    "MAX_NEWLINES": {
+      "COUNT": 20,
+      "PERIOD": 60
+    },
+    "PUNISHMENT": "none",
```
Comment (Tarek1337) [inline config/template.json]: Same as previously mentioned.

File: stage_1.txt [hunk 6]
```diff
@@ -0,0 +1 @@
+stage_1
```
Comment (Tarek1337) [inline stage_1.txt]: What

File: stage_2.txt [hunk 7]
```diff
@@ -0,0 +1 @@
+stage_2
```
Comment (Tarek1337) [inline stage_2.txt]: What 2

File: GearBot/Cogs/AntiSpam.py [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+from discord.ext import commands
+from discord import Object
+from discord.guild import Guild
+from discord.message import Message
+from discord.member import Member
+from discord.channel import TextChannel
+from Util.SpamBucket import SpamBucket
+from Util import Configuration, InfractionUtils, GearbotLogging, Utils, Translator
+from database.DatabaseConnector import Infraction
+import re
```
Comment (BlackHoleFox) [inline GearBot/Cogs/AntiSpam.py]: Just a general question, why use Regex here? There could be a cleaner and possibly more performant way to handle this. I'm imagining if there's a large spam incident, running Regex searches constantly may hit performance counts.

**Pull request merged**